### PR TITLE
Add operator position to binary and unary AST expressions

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Add operator position to binary and unary expressions AST expressions
Add operator position to corresponding BIR instructions 
Related to #604, #605

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/bir/bir.bal [hunk 1]
```diff
@@ -216,9 +216,9 @@ public type IntArithmeticBinaryInsn readonly & record {|
     *InsnBase;
     INSN_INT_ARITHMETIC_BINARY name = INSN_INT_ARITHMETIC_BINARY;
     ArithmeticBinaryOp op;
+    Position opPos;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: For instructions, the field can continue to be called `pos`: we are only having one position per insn.

File: compiler/modules/front.syntax/ast.bal [hunk 2]
```diff
@@ -73,7 +73,8 @@ public type CompoundAssignStmt record {|
     *PositionFields;
     LExpr lValue;
     Expr expr;
-    BinaryArithmeticOp|BinaryBitwiseOp op; 
+    BinaryArithmeticOp|BinaryBitwiseOp op;
+    Position opPos;
     Position pos;
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: Why do we need `pos` in addition to opPos, startPos and endPos?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/bir/bir.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: fixed

File: compiler/modules/front.syntax/ast.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal]: removed `pos` (replaced with opPos)

File: compiler/modules/bir/verify.bal [hunk 3]
```diff
@@ -330,7 +330,7 @@ function verifyCompareOperandTypeArray(VerifyContext vc, CompareInsn insn, t:Sem
 function verifyCompareOperandTypeBase(VerifyContext vc, CompareInsn insn, t:SemType operandType,
                                       t:UniformTypeBitSet expectType) returns err:Semantic?{
     if !t:isSubtypeSimple(operandType, expectType) {
-        return vc.err(`operand of ${insn.op} does not match order type`);
+        return vc.err(`operand of ${insn.op} at ${insn.pos} does not match order type`);
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: pos should be passed as a parameter

File: compiler/modules/bir/verify.bal [hunk 4]
```diff
@@ -342,7 +342,7 @@ function verifyEquality(VerifyContext vc, EqualityInsn insn) returns err:Semanti
             t:SemType intersectType = t:intersect(lhs.semType, rhs.semType);
             if !vc.isEmpty(intersectType) {
                 if insn.op.length() == 2 && !vc.isAnydata(lhs.semType) && !vc.isAnydata(rhs.semType) {
-                    return vc.err("at least one operand of an == or != expression must be a subtype of anydata");
+                    return vc.err(`at least one operand of an == or !=  at ${insn.pos} expression must be a subtype of anydata`);
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: pos should be passed as a parameter

File: compiler/modules/bir/verify.bal [hunk 5]
```diff
@@ -359,7 +359,7 @@ function verifyEquality(VerifyContext vc, EqualityInsn insn) returns err:Semanti
     else if isEqual(lhs, rhs) {
         return;
     }
-    return vc.err(`intersection of operands of operator ${insn.op} is empty`);
+    return vc.err(`intersection of operands of operator ${insn.op} at ${insn.pos} is empty`);
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: pos should be passed as a parameter

File: compiler/modules/front/codeGen.bal [hunk 6]
```diff
@@ -1189,17 +1209,17 @@ function codeGenLExprMappingKey(CodeGenContext cx, bir:BasicBlock block, Environ
     }
 }
 
-function codeGenArithmeticBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, bir:ArithmeticBinaryOp op, bir:Operand lhs, bir:Operand rhs, bir:Position pos) returns CodeGenError|ExprEffect {
+function codeGenArithmeticBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, bir:ArithmeticBinaryOp op, bir:Position opPos, bir:Operand lhs, bir:Operand rhs) returns CodeGenError|ExprEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Call the parameter `pos`.

File: compiler/modules/front/codeGen.bal [hunk 7]
```diff
@@ -1208,17 +1228,17 @@ function codeGenArithmeticBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, bir:A
...
     }
     else {
-        return cx.semanticErr(`${op} not supported for operand types`);
+        return cx.semanticErr(`${op} at ${opPos} not supported for operand types`);
     } 
     return { result, block: bb };
 }
 
-function codeGenBitwiseBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, s:BinaryBitwiseOp op, bir:IntOperand lhs, bir:IntOperand rhs) returns CodeGenError|ExprEffect {
+function codeGenBitwiseBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, s:BinaryBitwiseOp op, err:Position opPos, bir:IntOperand lhs, bir:IntOperand rhs) returns CodeGenError|ExprEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Call the parameter `pos`

File: compiler/modules/front/codeGen.bal [hunk 8]
```diff
@@ -1292,12 +1312,12 @@ function codeGenConstValue(CodeGenContext cx, bir:BasicBlock bb, Environment env
     }
 }
 
-function codeGenEquality(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:BinaryEqualityOp op, s:Expr left, s:Expr right) returns CodeGenError|ExprEffect {
+function codeGenEquality(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:BinaryEqualityOp op, err:Position opPos, s:Expr left, s:Expr right) returns CodeGenError|ExprEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Call the parameter `pos`

File: compiler/modules/front/codeGen.bal [hunk 9]
```diff
@@ -869,79 +870,98 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
-    var { result, block } = check codeGenCompoundableBinaryExpr(cx, block2, env, op, member, rexpr, pos);
+    var { result: member, block: block2 } = check codeGenMappingGet(cx, block1, mapping, "[", k, opPos);
+    var { result, block } = check codeGenCompoundableBinaryExpr(cx, block2, env, op, opPos, member, rexpr);
     bir:MappingSetInsn setInsn = { operands:[ mapping, k, result], position: lValue.pos };
     block.insns.push(setInsn);
     return { block };
 }
 
-function codeGenCompoundableBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:BinaryArithmeticOp|s:BinaryBitwiseOp op, bir:Register member, s:Expr rexpr, err:Position pos) returns CodeGenError|ExprEffect {
+function codeGenCompoundableBinaryExpr(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:BinaryArithmeticOp|s:BinaryBitwiseOp op, bir:Position opPos, bir:Register member, s:Expr rexpr) returns CodeGenError|ExprEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Call the parameter `pos`

File: compiler/modules/front/codeGen.bal [hunk 10]
```diff
@@ -869,79 +870,98 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
-function codeGenCompoundAssignToMappingMember(CodeGenContext cx, bir:BasicBlock bb, Environment env,
-                                              s:MemberAccessLExpr|s:FieldAccessLExpr lValue, bir:Register mapping, s:Expr rexpr, s:BinaryArithmeticOp|s:BinaryBitwiseOp op, err:Position pos) returns CodeGenError|StmtEffect {
+function codeGenCompoundAssignToMappingMember(CodeGenContext cx,
+                                              bir:BasicBlock bb,
+                                              Environment env,
+                                              s:MemberAccessLExpr|s:FieldAccessLExpr lValue,
+                                              bir:Register mapping,
+                                              s:Expr rexpr,
+                                              s:BinaryArithmeticOp|s:BinaryBitwiseOp op,
+                                              err:Position opPos) returns CodeGenError|StmtEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Call the parameter `pos`

File: compiler/modules/front/codeGen.bal [hunk 11]
```diff
@@ -869,79 +870,98 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
...
 
-function codeGenCompoundAssignToListMember(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:MemberAccessLExpr lValue, bir:Register list, s:Expr rexpr, s:BinaryArithmeticOp|s:BinaryBitwiseOp op, err:Position pos) returns CodeGenError|StmtEffect {
+function codeGenCompoundAssignToListMember(CodeGenContext cx,
+                                           bir:BasicBlock bb,
+                                           Environment env,
+                                           s:MemberAccessLExpr lValue,
+                                           bir:Register list,
+                                           s:Expr rexpr,
+                                           s:BinaryArithmeticOp|s:BinaryBitwiseOp op,
+                                           err:Position opPos) returns CodeGenError|StmtEffect {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Call the parameter `pos`.

File: compiler/modules/front/codeGen.bal [hunk 12]
```diff
@@ -869,79 +870,98 @@ function codeGenAssignToMember(CodeGenContext cx, bir:BasicBlock startBlock, Env
 }
 
 function codeGenCompoundAssignStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environment env, s:CompoundAssignStmt stmt) returns CodeGenError|StmtEffect {
-    var { lValue, expr, op, pos } = stmt;
+    var { lValue, expr, op, opPos } = stmt;
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: I would do `pos: opPos` i.e. call the parameter `pos`.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 12 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 11 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 10 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 9 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 8 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/bir/verify.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/verify.bal]: fixed

File: compiler/modules/bir/verify.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/verify.bal]: fixed

File: compiler/modules/bir/verify.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/verify.bal]: fixed

File: compiler/modules/front/codeGen.bal [hunk 13]
```diff
@@ -392,15 +392,15 @@ function codeGenForeachStmt(CodeGenContext cx, bir:BasicBlock startBlock, Enviro
     bir:BranchInsn branchToLoopHead = { dest: loopHead.label };
     initLoopVar.insns.push(branchToLoopHead);
     bir:Register condition = cx.createRegister(t:BOOLEAN);
-    bir:CompareInsn compare = { op: "<", orderType: t:UT_INT, operands: [loopVar, upper], result: condition };
+    bir:CompareInsn compare = { op: "<", pos: stmt.range.upper.startPos, orderType: t:UT_INT, operands: [loopVar, upper], result: condition };
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: On reflection, I think it's better to make this position be the position of `..<`.

File: compiler/modules/front.syntax/parseStmt.bal [hunk 14]
```diff
@@ -242,14 +245,13 @@ function finishAssignStmt(Tokenizer tok, LExpr|WILDCARD lValue, Position startPo
     return stmt;
 }
 
-function parseCompoundAssignStmt(Tokenizer tok, LExpr lValue, CompoundAssignOp op, Position startPos) returns CompoundAssignStmt|err:Syntax {
+function parseCompoundAssignStmt(Tokenizer tok, LExpr lValue, CompoundAssignOp op, Position opPos, Position startPos) returns CompoundAssignStmt|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal]: startPos argument after opPos seems a weird ordering to me, given that opPos is after startPos.

File: compiler/modules/front.syntax/ast.bal [hunk 15]
```diff
@@ -106,6 +106,7 @@ public type MatchStmt record {|
 public type MatchClause record {|
     MatchPattern[] patterns;
     Stmt[] block;
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal:109]: We should have startPos and endPos here also.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 13 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front.syntax/parseStmt.bal (hunk 14 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/parseStmt.bal]: fixed

File: compiler/modules/front.syntax/ast.bal (hunk 15 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal:109]: fixed

**Review by jclark — APPROVED** [review]

**Pull request merged**